ports/common/src/iot_network_openssl.c

```diff
@@ -752,28 +766,53 @@ IotNetworkError_t IotNetworkOpenssl_SetReceiveCallback( IotNetworkConnection_t p
                                                         IotNetworkReceiveCallback_t receiveCallback,
                                                         void * pContext )
 {
+    IOT_FUNCTION_ENTRY( IotNetworkError_t, IOT_NETWORK_BAD_PARAMETER );
     int posixError = 0;
-    IotNetworkError_t status = IOT_NETWORK_SUCCESS;
 
-    /* Set the callback and parameter. */
-    pConnection->receiveCallback = receiveCallback;
-    pConnection->pReceiveContext = pContext;
+    /* The receive callback must be non-NULL) */
+    if( receiveCallback != NULL )
+    {
+        /* Set the callback and parameter. */
+        pConnection->receiveCallback = receiveCallback;
+        pConnection->pReceiveContext = pContext;
 
-    posixError = pthread_create( &pConnection->receiveThread,
-                                 NULL,
-                                 _networkReceiveThread,
-                                 pConnection );
+        posixError = pthread_create( &pConnection->receiveThread,
+                                     NULL,
+                                     _networkReceiveThread,
+                                     pConnection );
 
-    if( posixError != 0 )
-    {
-        IotLogError( "Failed to create socket %d network receive thread. errno=%d.",
-                     pConnection->socket,
-                     posixError );
-        status = IOT_NETWORK_SYSTEM_ERROR;
+        if( posixError != 0 )
+        {
+            IotLogError( "Failed to create socket %d network receive thread. errno=%d.",
+                         pConnection->socket,
+                         posixError );
+            status = IOT_NETWORK_SYSTEM_ERROR;
+        }
+        else
+        {
+            pConnection->receiveThreadCreated = true;
+            status = IOT_NETWORK_SUCCESS;
+        }
     }
-    else
+
+    IOT_FUNCTION_EXIT_NO_CLEANUP();
+}
+
+/*-----------------------------------------------------------*/
+
+IotNetworkError_t IotNetworkOpenssl_SetCloseCallback( IotNetworkConnection_t pConnection,
+                                                      IotNetworkCloseCallback_t closeCallback,
+                                                      void * pContext )
+{
+    IotNetworkError_t status = IOT_NETWORK_BAD_PARAMETER;
+
+    if( closeCallback != NULL )
```

Comment:
Should we check value of if (pConnection  != NULL) here  (and in IotNetworkOpenssl_SetReceiveCallback )?

Reply:
I didn't add those here since they hadn't been here before, but this is a fair point.  Also, `assert()` might be more appropriate for both of these checks.

Reply:
I'm pretty sure there will be many such places, which we have overlooked earlier. I just happen to notice.  I'm not sure if CBMC proofs are written for interface and if that should have caught the problem.